library/std/src/os/unix/net/ancillary.rs

```diff
@@ -429,6 +441,34 @@ impl<'a> SocketAncillary<'a> {
         self.buffer.len()
     }
 
+    /// Returns the raw ancillary data as byte slice.
+    #[unstable(feature = "unix_socket_ancillary_data", issue = "76915")]
+    pub fn data(&self) -> &[u8] {
+        unsafe { MaybeUninit::slice_assume_init_ref(&self.buffer[..self.length]) }
+    }
+
+    /// Returns the entire buffer, including unused capacity.
+    ///
+    /// Use [`data()`](Self::data) if you are only interested in the used portion of the buffer.
+    #[unstable(feature = "unix_socket_ancillary_data", issue = "76915")]
+    pub fn buffer(&self) -> &[MaybeUninit<u8>] {
+        self.buffer
+    }
+
+    /// Returns the entire buffer as mutable slice, including unused capacity.
+    ///
+    /// You should normally call [`set_len()`](Self::set_len)
+    /// after changing the contents of the buffer.
+    ///
+    /// # Safety
+    /// All data written to the buffer must be valid ancillary data for the target platform,
+    /// and you must call [`set_len()`](Self::set_len) after changing
+    /// the buffer contents to update the internal bookkeeping.
+    #[unstable(feature = "unix_socket_ancillary_data", issue = "76915")]
+    pub unsafe fn buffer_mut(&mut self) -> &mut [MaybeUninit<u8>] {
+        self.buffer
+    }
```

Comment:
Is the use-case appending additional messages to the buffer? In that case it might be more ergonomic to only return the tail beyond the current length.

Reply:
The use-case I have in mind is to pass the entire buffer to a syscall to let the kernel write into it. I wasn't really thinking of manually extending it with hand-crafted messages.

If we would want to support that, I would lean towards making a public (unsafe) version of `add_to_ancillary_data`.

Reply:
I see, I thought this was about handling message types not supported by std rather than handling ancilliary data in conjunction with other syscalls. Are there syscalls other than `recvmsg` that would write ancillary data? io_uring I guess?

Reply:
Possibly, although I'm not familiar with `io_uring` myself. It also helps when implementing unsupported socket types on crates.io. I'd love to use this in [`tokio-seqpacket`](https://crates.io/crates/tokio-seqpacket) for example.

Reply:
> I'd love to use this in tokio-seqpacket for example.

At least for that particular usecase you probably wouldn't need it. You could temporarily create a `ManuallyDrop<UnixDatagram>` socket and then call `send_with_ancillary` since that boils down to the same syscall. At least once a [way to pass flags is provided](https://github.com/rust-lang/rust/issues/76915#issuecomment-820682675). That way a lot of code duplication can be avoided.

Reply:
I don't think that would be a good solution. That would mean abusing knowledge about implementation details of `std`. The API does not guarantee any mapping to specific syscalls or that `from_raw_fd` is pretty much a simple cast.

This proposed API change may be unsafe, but it does gives the required guarantees.